R/opt-cond-thread.R

```diff
@@ -0,0 +1,623 @@
+# @param node2 A numeric indicating the node ID of the second IF expression.
+#
+check_duplicate_expr <- function(fpd, node1, node2) {
+  first_expr <- fpd[fpd$id == node1, "text"]
+  second_expr <- fpd[fpd$id == node2, "text"]
+  
+  first_expr <- gsub(" ", "", first_expr, fixed = TRUE)
+  second_expr <- gsub(" ", "", second_expr, fixed = TRUE)
+  
+  return (first_expr == second_expr)
+}
+
+# Checks for exact negations in the greater than equal to conditions of the 
+# consecutive IF statements for the given node_id in the given fpd. 
+#
+# @param fpd A flatten parsed data data.frame.
+# @param node1 A numeric indicating the node ID of the first IF expression.
+# @param node2 A numeric indicating the node ID of the second IF expression.
+#
+check_comparsion_logic_ge <- function(fpd, node1, node2) {
+  first_expr <- fpd[fpd$id == node1, "text"]
+  second_expr <- fpd[fpd$id == node2, "text"]
+  if(length(grep(">=", first_expr)) > 0) {
+    first_expr <- gsub(" ", "", first_expr, fixed = T)
+    first_expr <- gsub(">=", "<", first_expr)
+    second_expr <- gsub(" ", "", second_expr, fixed = TRUE)
+    return (first_expr == second_expr)
+  } 
+  else if(length(grep(">=", second_expr)) > 0) {
+    second_expr <- gsub(" ", "", second_expr, fixed = TRUE)
+    first_expr <- gsub(" ", "", first_expr, fixed = TRUE)
+    second_expr <- gsub(">=", "<", second_expr)
+    return (first_expr == second_expr)
+  }
+  else{
+    return (FALSE)
+  }
+}
+
+# Checks for exact negations in the lesser than equal to conditions of the 
+# consecutive IF statements for the given node_id in the given fpd. 
+#
+# @param fpd A flatten parsed data data.frame.
+# @param node1 A numeric indicating the node ID of the first IF expression.
+# @param node2 A numeric indicating the node ID of the second IF expression.
+#
+check_comparsion_logic_le <- function(fpd, node1, node2) {
+  first_expr <- fpd[fpd$id == node1, "text"]
+  second_expr <- fpd[fpd$id == node2, "text"]
+  if(length(grep("<=", first_expr)) > 0) {
+    first_expr <- gsub(" ", "", first_expr, fixed = TRUE)
+    first_expr <- gsub("<=", ">", first_expr)
+    second_expr <- gsub(" ", "", second_expr, fixed = TRUE)
+    return (first_expr == second_expr)
+  } 
+  else if(length(grep("<=", second_expr)) > 0) {
+    first_expr <- gsub(" ", "", first_expr, fixed = TRUE)
+    second_expr <- gsub(" ", "", second_expr, fixed = TRUE)
+    second_expr <- gsub("<=", ">", second_expr)
+    return (first_expr == second_expr)
+  }
+  else{
+    return (FALSE)
+  }
+}
+
+# Checks for function calls in the IF statement conditions
+#
+# @param fpd A flatten parsed data.frame.
+# @param node1 A numeric indicating the node ID of the first IF expression.
+# @param node2 A numeric indicating the node ID of the second IF expression.
+#
+has_func_calls <- function(fpd, node1, node2) {
+  return ("SYMBOL_FUNCTION_CALL" %in% get_children(fpd, node1)$token | 
+            "SYMBOL_FUNCTION_CALL" %in% get_children(fpd, node2))  
```

Comment:
Is this going to compare in the whole data.frame?
Try always to use `&&` or `||`, unless you explicitly want to do vectors pointwise comparisons.

Reply:
Umm, actually I initially thought that this would be comparison between two booleans so I thought I could use both. But, I missed the  case where there are more than one function calls in a single `IF` statement condition. There this function would definitely throw warning or some unexpected behaviour. Thanks for this tip >.<